### PR TITLE
Turbocharge by using Jacobi representation, 2-ary NAF and precomputation

### LINKED ISSUE #94: Use more efficient algorithm for scalar multiplication
the current algorithm:
https://github.com/warner/python-ecdsa/blob/ffe9a61390c76ff6f6636e29f26b45c900535aaf/src/ecdsa/ellipticcurve.py#L141-L148

uses a simple double-and-add algorithm

try speeding it up by using either a lookup table or a sliding-window method

see also:
 * Wikipedia's [Elliptic curve point multiplication](https://en.wikipedia.org/wiki/Elliptic_curve_point_multiplication) and [Table of costs of operations in elliptic curves](https://en.wikipedia.org/wiki/Table_of_costs_of_operations_in_elliptic_curves)
 * [Scalar-multiplication algorithms](https://cryptojedi.org/peter/data/eccss-20130911b.pdf) presentation from Peter Schwabe
 * [Analysis of Efficient Techniques for Fast Elliptic Curve Cryptography  on x86-64 based Processors](https://eprint.iacr.org/2010/335.pdf) by Patrick Longa and Catherine Gebotys 
 * [Efficient and Secure ECDSA Algorithm and its Applications:  A Survey](https://arxiv.org/pdf/1902.10313.pdf) by Mishall Al-Zubaidie, Zhongwei Zhang and Ji Zhang

### PR DESCRIPTION
**Pull request by tomato42**
Use Jacobi representation for internal calculations.  Precompute point doublings for curve generators and public keys. Use 2-ary Non-Adjacent Form for multiplication (both for regular and precomputed points).

ver 0.14.1:
```
                  siglen    keygen   keygen/s      sign     sign/s    verify   verify/s
        NIST192p:     48   0.01550s     64.50   0.00826s    121.04   0.01605s     62.32
        NIST224p:     56   0.02100s     47.61   0.01131s     88.44   0.02281s     43.85
        NIST256p:     64   0.02828s     35.36   0.01514s     66.04   0.02988s     33.47
        NIST384p:     96   0.06678s     14.97   0.03536s     28.28   0.07114s     14.06
        NIST521p:    132   0.13108s      7.63   0.07089s     14.11   0.13921s      7.18
       SECP256k1:     64   0.02723s     36.72   0.01472s     67.93   0.02901s     34.47
```

this branch:
```
                  siglen    keygen   keygen/s      sign     sign/s    verify   verify/s
        NIST192p:     48   0.00033s   3062.43   0.00036s   2748.08   0.00062s   1605.66
        NIST224p:     56   0.00041s   2424.07   0.00046s   2196.71   0.00083s   1205.81
        NIST256p:     64   0.00053s   1892.05   0.00058s   1735.23   0.00106s    944.82
        NIST384p:     96   0.00110s    904.98   0.00118s    847.82   0.00217s    460.26
        NIST521p:    132   0.00234s    428.24   0.00245s    408.54   0.00443s    225.95
       SECP256k1:     64   0.00053s   1891.93   0.00058s   1734.46   0.00109s    913.35
 BRAINPOOLP160r1:     40   0.00025s   3982.49   0.00029s   3490.15   0.00053s   1878.51
 BRAINPOOLP192r1:     48   0.00032s   3086.07   0.00036s   2761.68   0.00063s   1578.22
 BRAINPOOLP224r1:     56   0.00041s   2412.41   0.00046s   2185.52   0.00076s   1311.65
 BRAINPOOLP256r1:     64   0.00054s   1866.84   0.00058s   1719.45   0.00110s    906.85
 BRAINPOOLP320r1:     80   0.00077s   1306.97   0.00083s   1201.59   0.00158s    632.82
 BRAINPOOLP384r1:     96   0.00112s    892.44   0.00119s    841.48   0.00229s    436.71
 BRAINPOOLP512r1:    128   0.00214s    467.05   0.00226s    441.64   0.00422s    237.13
```

for comparison, OpenSSL on same machine achieves:
```
                              sign    verify    sign/s verify/s
 256 bits ecdsa (nistp256)   0.0000s   0.0001s  45320.1  14171.3
 384 bits ecdsa (nistp384)   0.0008s   0.0006s   1259.5   1653.3
 256 bits ecdsa (brainpoolP256r1)   0.0003s   0.0003s   2983.5   3333.2
 384 bits ecdsa (brainpoolP384r1)   0.0008s   0.0007s   1258.8   1528.1
 512 bits ecdsa (brainpoolP512r1)   0.0015s   0.0012s    675.1    860.1
```
(OpenSSL has hand-optimised assembly with AVX for `nistp256`/`NIST256p` curve)

depends on #126, fixes #94 (while there are more efficient algorithms still, being 3.6 times slower than generic C implementation for verification and 33% slower for signing is acceptable to me :smile: )

or when gmpy2 is used, being like 7% slower for verify and more than 2 times faster for signing :grin:
```
                  siglen    keygen   keygen/s      sign     sign/s    verify   verify/s
        NIST192p:     48   0.00016s   6180.12   0.00017s   5846.30   0.00033s   3029.51
        NIST224p:     56   0.00021s   4861.86   0.00021s   4662.63   0.00042s   2366.47
        NIST256p:     64   0.00023s   4343.93   0.00024s   4152.79   0.00047s   2148.83
        NIST384p:     96   0.00040s   2507.97   0.00041s   2435.99   0.00079s   1260.01
        NIST521p:    132   0.00088s   1135.13   0.00089s   1121.94   0.00139s    721.07
       SECP256k1:     64   0.00023s   4360.83   0.00024s   4147.61   0.00044s   2254.53
 BRAINPOOLP160r1:     40   0.00014s   7261.37   0.00015s   6824.47   0.00031s   3248.21
 BRAINPOOLP192r1:     48   0.00016s   6219.18   0.00017s   5862.93   0.00034s   2933.74
 BRAINPOOLP224r1:     56   0.00021s   4876.48   0.00022s   4640.40   0.00041s   2413.48
 BRAINPOOLP256r1:     64   0.00023s   4397.89   0.00024s   4178.48   0.00044s   2272.76
 BRAINPOOLP320r1:     80   0.00031s   3246.64   0.00032s   3138.38   0.00063s   1593.14
 BRAINPOOLP384r1:     96   0.00040s   2491.04   0.00041s   2421.67   0.00079s   1262.64
 BRAINPOOLP512r1:    128   0.00062s   1618.30   0.00063s   1577.42   0.00125s    799.29
```

To do:
 - [x] rename the "classical" representation to "affine"
 - [x] merge fixup commits to earlier ones
 - [x] add documentation
 - [x] test coverage
 - [x] make sure that all the expensive test cases use the new PointJacobi implementation
 - [x] use jacobi representation for Brainpool curves too
 - [x] avoid constructing PointJacobi objects over and over when multiplying
 - [x] add to travis tests with oldest `gmpy` and `gmpy2` that work for us, specify them in `setup.py`

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -53,31 +79,55 @@ On an Intel Core i7 4790K @ 4.0GHz I'm getting the following performance:
...
+ BRAINPOOLP384r1:     96   0.00040s   2491.04   0.00041s   2421.67   0.00079s   1262.64
+ BRAINPOOLP512r1:    128   0.00062s   1618.30   0.00063s   1577.42   0.00125s    799.29
+```
+
+(there's also `gmpy` version, execute it using `tox -e speedgmpy`)
 
 For comparison, a highly optimised implementation (including curve-specific
-assemply) like OpenSSL provides following performance numbers on the same
-machine. Run `openssl speed` to reproduce it:
+assemply for some curve), like the one in OpenSSL, provides following
```
Comment (t8m) [inline README.md]: Nit: assembly

File: src/ecdsa/ellipticcurve.py [hunk 2]
```diff
@@ -63,26 +102,482 @@ def a(self):
...
+      Z3 = Z1*(X2-X1) %p
+      return X3, Y3, Z3
+
+  def _add_with_z2_1(self, X1, Y1, Z1, X2, Y2, p):
+      """add points when Z2 == 1"""
+      # after:
+      # http://hyperelliptic.org/EFD/g1p/auto-shortw-jacobian.html#addition-madd-2007-bl
+      Z1Z1 = Z1*Z1 % p
+      U2, S2 = X2*Z1Z1%p, Y2*Z1*Z1Z1%p
+      H = U2-X1%p
```
[the remark this answers is not in the document]
Comment (tomato42) [reply src/ecdsa/ellipticcurve.py]: ah, good catch

File: src/ecdsa/ellipticcurve.py [hunk 3]
```diff
@@ -63,26 +102,482 @@ def a(self):
...
+      return X3, Y3, Z3
+
+  def _add_with_z_eq(self, X1, Y1, Z1, X2, Y2, p):
+      """add points when Z1 == Z2"""
+      # after:
+      # http://hyperelliptic.org/EFD/g1p/auto-shortw-jacobian.html#addition-zadd-2007-m
+      A = (X2-X1)**2%p
+      B = X1*A%p
+      C = X2*A
+      D = (Y2-Y1)**2 %p
```
[the remark this answers is not in the document]
Comment (tomato42) [reply src/ecdsa/ellipticcurve.py]: yes, it should be consistent

File: src/ecdsa/ellipticcurve.py [hunk 4]
```diff
@@ -63,26 +102,482 @@ def a(self):
...
+      Y3 = ((Y2-Y1)*(B-X3)-Y1*(C-B)) %p
+      Z3 = Z1*(X2-X1) %p
+      return X3, Y3, Z3
+
+  def _add_with_z2_1(self, X1, Y1, Z1, X2, Y2, p):
+      """add points when Z2 == 1"""
+      # after:
+      # http://hyperelliptic.org/EFD/g1p/auto-shortw-jacobian.html#addition-madd-2007-bl
+      Z1Z1 = Z1*Z1 % p
+      U2, S2 = X2*Z1Z1%p, Y2*Z1*Z1Z1%p
```
[the remark this answers is not in the document]
Comment (tomato42) [reply src/ecdsa/ellipticcurve.py]: this makes it faster

[the remark this answers is not in the document]
Comment (t8m) [reply src/ecdsa/ellipticcurve.py]: I suppose not significantly, but I am fine with it as is anyway.

File: src/ecdsa/test_jacobi.py [hunk 5]
```diff
@@ -0,0 +1,365 @@
...
+        self.assertEqual(generator_256, pw)
+
+    def test_single_double(self):
+        pj = PointJacobi.from_affine(generator_256)
+        pw = generator_256.double()
+
+        pj = pj.double()
+
+        self.assertEqual(pj.x(), pw.x())
+        self.assertEqual(pj.y(), pj.y())
```
[the remark this answers is not in the document]
Comment (tomato42) [reply src/ecdsa/test_jacobi.py]: fixed

File: src/ecdsa/test_jacobi.py [hunk 6]
```diff
@@ -0,0 +1,365 @@
...
+        self.assertIs(pj, INFINITY)
+
+    def test_multiply_by_one(self):
+        pj = PointJacobi.from_affine(generator_256)
+        pw = generator_256 * 1
+
+        pj = pj * 1
+
+        self.assertEqual(pj.x(), pw.x())
+        self.assertEqual(pj.y(), pj.y())
```
[the remark this answers is not in the document]
Comment (tomato42) [reply src/ecdsa/test_jacobi.py]: fixed

File: src/ecdsa/test_jacobi.py [hunk 7]
```diff
@@ -0,0 +1,365 @@
...
+
+    def test_multiply_by_two(self):
+        pj = PointJacobi.from_affine(generator_256)
+        pw = generator_256 * 2
+
+        pj = pj * 2
+
+        self.assertEqual(pj.x(), pw.x())
+        self.assertEqual(pj.y(), pj.y())
```
[the remark this answers is not in the document]
Comment (tomato42) [reply src/ecdsa/test_jacobi.py]: fixed

File: src/ecdsa/ellipticcurve.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (tomato42) [reply src/ecdsa/ellipticcurve.py]: yes, it's just like 20-40% faster, but this lives in a tight loop, so it adds up

File: README.md (hunk 1 above)
Comment (tomato42) [reply README.md]: and fixed

**Comment by tomato42** [comment]
@t8m ok, fixed and rebased

**Pull request merged**